### PR TITLE
TEN-120 update sign in again with SA credential page & controller

### PR DESCRIPTION
**Pull request by TSiyamachiraQA**

### CODE REVIEW

File: app/uk/gov/hmrc/taxenrolmentassignmentfrontend/controllers/SignInAgainController.scala [hunk 1]
```diff
@@ -16,33 +16,71 @@
...
+import uk.gov.hmrc.taxenrolmentassignmentfrontend.repository.TEASessionCache
+import uk.gov.hmrc.taxenrolmentassignmentfrontend.services.{EACDService, UsersGroupSearchService}
 import uk.gov.hmrc.taxenrolmentassignmentfrontend.views.html.SignInAgain
 
 import javax.inject.{Inject, Singleton}
-import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.ExecutionContext
 
 @Singleton
 class SignInAgainController @Inject()(
```
Comment (Chrissystk) [inline app/uk/gov/hmrc/taxenrolmentassignmentfrontend/controllers/SignInAgainController.scala:35]: I think that this signout function is still need as there are links of other pages to signout (i.e PTOnAnotherAccount).  I think what you have done needs to be in a new controller specific for SigninAgainWithSAAccount

**Review by Chrissystk — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/taxenrolmentassignmentfrontend/controllers/SignInAgainController.scala (hunk 1 above)
Comment (Chrissystk) [reply app/uk/gov/hmrc/taxenrolmentassignmentfrontend/controllers/SignInAgainController.scala:35]: Sorry ignore that just realised it is the SignOutController that is called in other places so overwriting that controller is fine just maybe rename it :-)

Comment (TSiyamachiraQA) [reply app/uk/gov/hmrc/taxenrolmentassignmentfrontend/controllers/SignInAgainController.scala]: I made changes to the controller and view name to be along the lines of `SignInWithSAAccount`

File: app/uk/gov/hmrc/taxenrolmentassignmentfrontend/controllers/SignInWithSAAccountController.scala [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+          usersGroupSearchService
+            .getAccountDetails(response.enrolledCredential.get)
+            .value
+            .map {
+              case Right(saAccountDetails) =>
+                  Ok(
+                    signInWithSAAccountPage(
+                      saAccountDetails,
+                      signOutController,
+                      reportSuspiciousIDController
```
Comment (Chrissystk) [inline app/uk/gov/hmrc/taxenrolmentassignmentfrontend/controllers/SignInWithSAAccountController.scala:66]: You don't need to pass controllers into the view you can access them already in the view.

File: app/uk/gov/hmrc/taxenrolmentassignmentfrontend/views/SignInWithSAAccount.scala.html [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  @createSummaryListRow(messages("signInAgain.email"), saAccountDetails.email.get)
+  }
+  @createSummaryListRow(messages("signInAgain.lastLoginDate"), saAccountDetails.lastLoginDate)
+  @for(mfaDetail <- saAccountDetails.mfaDetails) {
+  @createSummaryListRow(mfaDetail.factorName, mfaDetail.factorValue)
+  }
+ </dl>
+
+ <p class="govuk-body">
+   <a class="govuk-link" id="reportId" href="@reportSuspiciousIDController.view()">
```
Comment (Chrissystk) [inline app/uk/gov/hmrc/taxenrolmentassignmentfrontend/views/SignInWithSAAccount.scala.html:59]: href = "@routes.ReportSuspiciousIDController.view" can be used just make sure your import uk.gov.hmrc.taxenrolmentassignmentfrontend.controllers.routes

File: app/uk/gov/hmrc/taxenrolmentassignmentfrontend/views/SignInWithSAAccount.scala.html [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    @messages("signInAgain.linkText")
+   </a>
+ </p>
+<!--//TODO remove fraud reporting url after consultation with leads, as it requires other changes such remove couple of tests etc etc-->
+
+  @govukButton(Button(
```
Comment (Chrissystk) [inline app/uk/gov/hmrc/taxenrolmentassignmentfrontend/views/SignInWithSAAccount.scala.html:65]: This should be a post. See LandingPageView for example

File: conf/app.routes [hunk 5]
```diff
@@ -8,7 +8,7 @@ GET        /enrol-pt/enrol-current-user-id  uk.gov.hmrc.taxenrolmentassignmentfr
 GET        /enrol-pt/contact-hmrc-sa  uk.gov.hmrc.taxenrolmentassignmentfrontend.controllers.ReportSuspiciousIDController.view
 GET        /enrol-pt/enrolment-success      uk.gov.hmrc.taxenrolmentassignmentfrontend.controllers.CurrentIdConfirmationController.view
 GET        /logout                  uk.gov.hmrc.taxenrolmentassignmentfrontend.controllers.SignOutController.signOut()
-GET        /enrol-pt/sign-in-again          uk.gov.hmrc.taxenrolmentassignmentfrontend.controllers.SignInAgainController.view
+GET        /enrol-pt/sign-in-again          uk.gov.hmrc.taxenrolmentassignmentfrontend.controllers.SignInWithSAAccountController.view
 GET        /pt-enrolment-other-account   uk.gov.hmrc.taxenrolmentassignmentfrontend.controllers.PTEnrolmentOnOtherAccountController.view
```
Comment (Chrissystk) [inline conf/app.routes:12]: Need to have a POST here too

File: app/uk/gov/hmrc/taxenrolmentassignmentfrontend/controllers/SignInWithSAAccountController.scala [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+          )
+          sessionCache.getEntry[String](REDIRECT_URL).map {
+            case Some(redirectUrl) =>
+              Redirect(routes.AccountCheckController.accountCheck(redirectUrl))
+            case None =>
+              InternalServerError
+          }
+      }
+
+  }
```
Comment (Chrissystk) [inline app/uk/gov/hmrc/taxenrolmentassignmentfrontend/controllers/SignInWithSAAccountController.scala:85]: A POST action needs to be put in here. If you look at the LandingPage you will see how it is done :-)

**Review by Chrissystk — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/taxenrolmentassignmentfrontend/controllers/SignInWithSAAccountController.scala (hunk 6 above)
Comment (TSiyamachiraQA) [reply app/uk/gov/hmrc/taxenrolmentassignmentfrontend/controllers/SignInWithSAAccountController.scala:90]: I've removed the controllers I was passing in and I have now introduced routes in the view. I also added the POST action for the button and its respective route. 👍

File: app/uk/gov/hmrc/taxenrolmentassignmentfrontend/controllers/SignInWithSAAccountController.scala [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    }
+  }
+
+  def continue: Action[AnyContent] = authAction.async { implicit request =>
+    sessionCache.getEntry[String](REDIRECT_URL).map {
+      case Some(redirectUrl) =>
+        Redirect(redirectUrl)
+      case None => InternalServerError
+    }
+  }
```
Comment (Chrissystk) [inline app/uk/gov/hmrc/taxenrolmentassignmentfrontend/controllers/SignInWithSAAccountController.scala:73]: Can we change this to match the other controller (logging and using errorHandler so the error page will be displayed). This is the function
`  def continue: Action[AnyContent] = authAction.async { implicit request =>
    sessionCache.getEntry[String](REDIRECT_URL).map {
      case Some(redirectUrl) =>
        logger.logEvent(
          logRedirectingToReturnUrl(
            request.userDetails.credId,
            "[EnrolledForPTController][continue]"
          )
        )
        Redirect(redirectUrl)
      case None =>
        handleErrors(
          NoRedirectUrlInCache,
          "[EnrolledForPTController][continue]"
        )
    }
  }
`

**Review by Chrissystk — CHANGES_REQUESTED** [review]

**Review by Chrissystk — APPROVED** [review]

**Pull request merged**